### PR TITLE
Limit concurrency to avoid hitting browser limits

### PR DESCRIPTION
**Pull request by ironicnet**
When trying to download the amazing models from https://mikelovesrobots.github.io/mmmm/ i got the following error:
image

After researching, i found out it was due to hitting the api too frequently.

Sooo, i added a semaphore to ensure we don't hit that limit.
Currently will process 25 concurrent downloads. This can be tuned by setting the `concurrent` parameter in the url.

Tested it locally and i had no trouble downloading a directory using this parameters: `?url=https%3A%2F%2Fgithub.com%2Fmikelovesrobots%2Fmmmm%2Ftree%2Fmaster%2Funity-plugin&concurrent=20`

### CODE REVIEW

**Comment by ironicnet** [comment]
I tested using 100 concurrent files and it worked flawlessly. Not sure what's the right number

**Comment by fregante** [comment]
Thanks for the PR! I'm not sure a plain limit like that makes sense. If the server can physically respond to 60 requests in 1 minute, then even if we limit it to 20 concurrent requests it'll still resolve 20 requests every 20 seconds. Maybe it only slows it down a bit but I think the issue is amount of request per minute rather than concurrency. 

Maybe this, if it already helps, can be coupled with `p-retry`, in order to *also* automatically retry failed requests. 

I suggest using `p-map` and `p-retry` rather than this semaphore implementation, you'll be able to add them both *outside* the download function.

**Comment by fregante** [comment]
Also something is up with the diff because there are thousands of lines changed in the lockfile. Ensure you start fresh and avoid unnecessary changes when installing dependencies.

**Comment by ironicnet** [comment]
Sorry, my wording was wrong. The error happens at browser level, not at server level. Is the browser which runs out of resources to fetch so many urls. The requests never hits the server, because the browser cannot handle it.

I'll check the package-lock.json, sorry, i didn't notice that change. (I probably must have done a `npm i` instead of a `npm ci`).

Also, i will check how this can work with the `p-retry` and `p-map`

**Comment by fregante** [comment]
> The error happens at browser level

Wow, indeed. I wonder if the browser can even zip that many files once they've been downloaded. Were you able to download them all after this change?

**Comment by ironicnet** [comment]
> > The error happens at browser level
> 
> Wow, indeed. I wonder if the browser can even zip that many files once they've been downloaded. Were you able to download them all after this change?

Yeap, i had no trouble zipping and it worked great. At least with 3500 files (15 mb) it worked great.
My pc has 16gb ram, so zipping files in memory should be an issue...

**Comment by fregante** [comment]
Do you have time to complete this?

**Comment by ironicnet** [comment]
Actually i have some free time, i will tackle this this days

**Comment by ironicnet** [comment]
@fregante It seems that the package-lock is not updated. Doing a fresh install and then a simple `npm install` already brings big changes into the `package-lock.json`.

I can create a PR to just sync the `package-lock.json` and then update this branch

**Comment by fregante** [comment]
I don’t think the lockfile is the problem, it's just that this PR deletes it. You should be able to just work on the version from the main branch without lockfile changes.

**Comment by ironicnet** [comment]
Ok, i didn't get to read your last comment and i created this PR. feel free to close it: #53

**Comment by fregante** [comment]
Hi, we talked about using p-retry since it has better retry handling on network failure as well

**Review by fregante — COMMENTED** [review]
Looks good so far!

CI is failing likely due to an old Parcel version. Let me see what I can do

**Comment by fregante** [comment]
It turns out I'm still using Parcel v1 on this project, which doesn't support new syntax. I downgraded a dependency so we don't have this issue.

After this PR is merged I'll look into updating Parcel (PR welcome too)

File: index.js [hunk 1]
```diff
@@ -94,6 +99,7 @@ async function init() {
 
 	try {
 		const query = new URLSearchParams(location.search);
+		concurrentDownloads = Number.parseInt(query.get('concurrent') || DEFAULT_CONCURRENT_DOWNLOADS, 10);
```
[the remark this answers is not in the document]
Comment (ironicnet) [reply index.js]: Removed the option from the query.

File: index.js [hunk 2]
```diff
@@ -159,11 +165,16 @@ async function init() {
 	};
 
 	let downloaded = 0;
+	const downloadFile = async file => pRetry(() => repoIsPrivate ? fetchPrivateFile(file) : fetchPublicFile(file),
```
[the remark this answers is not in the document]
Comment (ironicnet) [reply index.js]: I knew there was a smell in that syntax, but didn't realize where. Thanks for pointing it out.

File: index.js [hunk 3]
```diff
@@ -159,11 +165,16 @@ async function init() {
...
+		{
+			onFailedAttempt: async error => {
+				await console.error(`Error downloading ${file.url}. Attempt ${error.attemptNumber}. ${error.retriesLeft} retries left.`);
+			},
+			retries: DEFAULT_RETRIES,
```
[the remark this answers is not in the document]
Comment (ironicnet) [reply index.js]: 10 retries is the default, i left it that way

**Comment by fregante** [comment]
Thank you! I verified both the issue and that this fixes it 🙌

**Pull request merged**